[runtime] Infrastructure to build runtime components using NativeAOT

lambdageek opened the pull request:
**Motivation**: While the .NET runtime itself is perfectly capable of running .NET code, we also provide other components that currently have to be implemented in native code.  For example the DAC - its job is to introspect a live CoreCLR process or a memory dump and provide the information to diagnostic tools.  The DAC is not a .NET runtime and it exposes a native interface to other tooling.  But as we evolve the DAC it might make sense to implement parts of it in C#.

![yodawg](https://github.com/dotnet/runtime/assets/480437/c47218cb-0897-4627-957a-adb36b907d13)

# Native runtime component libraries using NativeAOT

This  PR allows managed libraries to be compiled using NativeAOT and to be used in the runtime using normal CMake idioms.

## Adding a new managed library

Add a new subdirectory to `src/native/managed` for your library with a `src` and `inc` subdirectories:

``` console
$ mkdir -p libMyNewLibrary/src libMyNewLibrary/inc
$ dotnet new classlib -n libMyNewLibrary -o libMyNewLibrary/src
```

In `src/native/managed/compile-native.proj`, add
`src/native/managed/libMyNewLibrary/src/libMyNewLibrary.csproj` to the `NativeLibsProjectsToBuild`
item group.

In `src/native/managed/libMyNewLibrary/src/libMyNewLibrary.csproj`:
1. Near the top,  add `<Import Project="..\..\native-library.props" />`
2. Near the bottom, add `<Import Project="..\..\native-library.targets" />`
3. Define an item `@(InstallRuntimeComponentDest)` that has directory names relative to `artifacts/bin/<runtimeFlavor>/<os.arch.config>/` where the shared library should be installed.  It's a good idea to have at least `.`:
    ```xml
      <ItemGroup>
          <InstallRuntimeComponentDest Include="." />
          <InstallRuntimeComponentDest Include="sharedFramework" Condition="'$(RuntimeFlavor)' == 'coreclr'"/>
      </ItemGroup>
    ```

Limitations:

*  The project should be called `libXXXX` - currently the infrastructure expects a `lib` prefix on all platforms.
* Currently only shared library output is supported.  In principle static linking is possible, but the
infrastructure is not finished yet.  Additionally, mixing Debug/Release configurations with static
linking will not be supported on Windows.

--- 

Ideally we'd like to be able to produce shared or static libraries.  Currently only shared libraries really work robustly (in that only the exported native symbols of the C# library are exposed and the NativeAOT implementation details are completely hidden).

There was an attempt at doing static linking too, but it needs a bit more work on Linux/Mac (using `-r` to create relocatable objects in conjunction with `-Wl,--exclude-libs` or `-Wl,-hidden-l` to hide the symbols of the NativeAOT runtime).  Doing naïve static linking results in duplicate GC and eventpipe symbols (since CoreCLR and NativeAOT both define them).  Static linking on Windows probably isn't worthwhile due to #98356 and because I'm not sure if hiding symbols will be possible.

There was some work to have cmake integration: runtime could do `find_nativeaot_library(libWhatever [REQUIRED])` and get two cmake targets: `libWhatever::libs` and `libWhatever::headers` that could be used for linking with other targets.  The current usecase doesn't need this level of integration however, so the cmake stuff was moved to this branch: https://github.com/lambdageek/runtime/tree/naot-runtime-lib-with-cmake

lambdageek committed:
wip: dotnet.sh build src/native/libhellomanaged/compile-native.proj works

lambdageek committed:
hide unnecesary dependencies

lambdageek committed:
add compile-native.proj to runtime-prereqs.proj

lambdageek committed:
make first builds work

was missing dependency resolution for the native projects

lambdageek committed:
add an include dir for libhellomanaged

lambdageek committed:
wire in static lib support; use to call libhellomanaged from daccess

lambdageek committed:
call hellomanaged_Hello from DAC DllMain

lambdageek committed:
static linking of nativeAotFramework

lambdageek committed:
support for importing NativeAOT compiled shared libs

lambdageek committed:
build native libs as shared by default

lambdageek committed:
WIP: first try at installing the native lib

lambdageek committed:
relocatable linking for static libs

GNU ld and clang's lld support a '-r' options for linking a bunch of
object files together into another one.  Maybe we can use this
together with the exclude-libs/hidden-lx options to combine just the
native lib and the runtime support libs before doing the final link.

unclear if there's anything similar on Windows.

lambdageek committed:
cleanup shared lib version; add install_clr support

lambdageek committed:
rm TODO

lambdageek committed:
don't build hellomanaged on platforms without NativeAOT

also don't build for Mono for now

lambdageek committed:
fixup windows build

lambdageek committed:
fixup win implib

lambdageek committed:
disable more configurations

lambdageek committed:
try to workaround dotnet/msbuild#2811

lambdageek commented:
I don't understand if I'm hitting dotnet/msbuld#2811 or if I messed up something else

```
.dotnet\sdk\9.0.100-alpha.1.23615.4\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(266,5): error NETSDK1047: (NETCORE_ENGINEERING_TELEMETRY=Build) Assets file 'D:\a\_work\1\s\artifacts\obj\libhellomanaged\project.assets.json' doesn't have a target for 'net9.0/win-arm64'. Ensure that restore has run and that you have included 'net9.0' in the TargetFrameworks for your project. You may also need to include 'win-arm64' in your project's RuntimeIdentifiers.
```

What I want to happen is that before `runtime.proj` runs cmake, something else in our build should run NativeAOT on `libhellomanaged.csproj` .  So `compile-native.proj` just uses the MSBuild task to invoke `Publish` on `libhellomanaged.csproj`.   But just doing it naively, I get errors about not restoring `libhellomanaged.csproj`.  So I tries `Restore;Build;Publish` - which I think you're not supposed to do.  So then I changed it to two separate MSBuild task invocations.  But it seems like that's not enough either?

lambdageek committed:
allow daccess to keep building if hellomanaged is not present

lambdageek committed:
always use lld on linux?

lambdageek committed:
put AdditionalProperties on the ProjectReference same as msbuild publish

make sure the same runtime identifier is passed down

lambdageek commented:
Linux-arm64 builds aren't happy I guess because they're not looking at the right C toolchain since we're compiling those on x64 build machines.

lambdageek committed:
no x86 or riscv

lambdageek committed:
linux-arm64 hack

MichalStrehovsky commented:
One general concern I'd have around this is platform support - native AOT only supports a subset of platforms that CoreCLR/Mono can target. Would this put native AOT on a critical path in platform bringup? It's not necessarily a bad thing, but we have existing gaps like WASM/WASI with Mono or x86 Windows with CoreCLR-JIT (plus whatever community ports exist for each of the runtime).

lambdageek committed:
HACK: installer

lambdageek committed:
disable armel

lambdageek committed:
remove bfd hack - it didn't work

lambdageek committed:
pass SysRoot and --gcc-toolchain to native build

lambdageek committed:
fix typo (RID is linux-riscv64 not linux-riscv)

lambdageek committed:
set linker flavor if doing cross builds

lambdageek committed:
fix for cmake 3.20

lambdageek committed:
use GenerateFileFromTemplate for shared mode

lambdageek committed:
use GenerateFileFromTemplate for static libs

lambdageek committed:
rename toplevel dir to src/native/managed

lambdageek commented:
/azp run help

lambdageek commented:
/azp run

lambdageek commented:
/azp list

lambdageek commented:
/azp run runtime-coreclr outerloop,runtime-community,runtime-tools-tests,runtime-sanitized

lambdageek committed:
allow csproj to specify cmake include dir for native library

lambdageek committed:
fixup INCLUDE path for Windows

lambdageek committed:
Use find_package to find native AOT libs

lambdageek committed:
Add README

lambdageek committed:
remove unused function

lambdageek committed:
libhellomanaged is optional

lambdageek committed:
move user-facing cmake to eng/functions.cmake

lambdageek committed:
remove cruft

lambdageek committed:
fix windows paths

lambdageek committed:
Use unix-style paths even on Windows; cmake prefers this

lambdageek committed:
dependency tracking and stripping

lambdageek commented:
I think this is almost ready - working on dependency tracking and symbol stripping.  (They turn out to be the same thing, basically - we need a custom command and custom target to track when the nativeaot output has become out of date; if it has, we also have a chance to add a post_build event to the copy target that will do the symbol stripping)

jkotas commented:
This is duplicating most of https://github.com/dotnet/runtime/tree/main/src/coreclr/nativeaot/BuildIntegration in cmake. If you are only supporting shared library output, do we need to have all this duplication? Can you just use `dotnet publish` of the library and reference the output of `dotnet publish` command instead?

lambdageek commented:
> This is duplicating most of https://github.com/dotnet/runtime/tree/main/src/coreclr/nativeaot/BuildIntegration in cmake. If you are only supporting shared library output, do we need to have all this duplication? Can you just use `dotnet publish` of the library and reference the output of `dotnet publish` command instead?

The only bits that are duplicating BuildIntegration are the static library support in :

https://github.com/dotnet/runtime/pull/98565/files#diff-50f0447f5f00a794a431f0ece5d8b308442ca22079f8e8a2eb97e84e88c36622R118-R159

and the `add_nativeAotFramework_targets_once` function.

If I drop static library support (which doesn't work anyway), those will be the only bits that get removed (I'm not opposed).

The rest is just infrastructure so you can do the following two things:

1.  Add a classlib under `src/native/managed` and then `runtime-prereqs.proj` will in effect invoke `dotnet publish` on your library (that's what `compile-native.proj` is responsible for):

2. You can say the following in cmake and the right things will happen:

   ```cmake
     find_nativeaot_library (libMyHelperLibrary REQUIRED)
   
      ...
     target_link_libraries (mscordac PRIVATE libMyHelperLibrary::libMyHelperLibrary)
     ...
     install_clr (TARGETS libMyHelperLibrary::libMyHelperLibrary DESTINATION ...)
   ```

None of these last two bits duplicate anything from BuildIntegration.  They are built on top of `dotnet publish -p:PublishAot=true`

lambdageek committed:
windows stripping support

lambdageek committed:
cleanup

lambdageek committed:
update README

lambdageek committed:
remove FIXME

lambdageek committed:
remove static linking support

it wasn't a good fit for CoreCLR because the EventPipe and GC symbols
clashed between CoreCLR and NativeAOT during static linking.

A possible way forward is to use incremental linking (`clang -Wl,-r`) on Linux/Mac in
combination with the `--exclude-libs` or `-hidden-lx` linker options.

However on Windows since NativeAOT static artifacts are always built
in Release mode, we would only be able to do static linking in limited
circumstances (and we would need to find the right linker option to do
the apropriate symbol hiding / incremental linking)

jkoritzinsky commented on eng/native/import-nativeaot-library.cmake:
I think in the future, I'd almost rather have a "CMake module" file produced for the NativeAOT libraries instead of hard-coding it all here. I opened an issue related to this in DNNE a while back, https://github.com/AaronRobinsonMSFT/DNNE/issues/185.

lambdageek replied:
yea, this could be a separate NuGet.  it should be possible to generate a cmake config, a rust crate, etc just by hooking something up after NativeAOT's publish target.

right now the cmake  config isn't entirely self-contained most obviously it depends on eng/native/configureplatform.cmake but probably also some other of our variables are leaking in.  But maybe it's possible to refactor it.

Also I didn't do multi-config generator support, which maybe people care about.

jkoritzinsky commented on eng/native/functions.cmake:
```diff
@@ -367,26 +367,54 @@ function(generate_exports_file_prefix inputFilename outputFilename prefix)
+    else()
+      get_property(strip_source_file TARGET ${targetName} PROPERTY IMPORTED_LOCATION)
+      # for an imported library, we will hang the post-build event on the copy target
+      get_property(copy_target TARGET ${targetName} PROPERTY CLR_IMPORTED_NATIVEAOT_LIBRARY_COPY_TARGET)
```
Could we rename `CLR_IMPORTED_NATIVEAOT_LIBRARY_COPY_TARGET` to `CLR_IMPORTED_COPY_TARGET` and unconditionally check for it instead of conditioning it on `CLR_IMPORTED_NATIVEAOT_LIBRARY`?
And for the other cases, could we use the `IMPORTED` target property?

lambdageek replied:
> And for the other cases, could we use the `IMPORTED` target property?

I didn't want to prescribe behavior for other imported libraries

> Could we rename CLR_IMPORTED_NATIVEAOT_LIBRARY_COPY_TARGET to CLR_IMPORTED_COPY_TARGET and unconditionally check for it instead of conditioning it on CLR_IMPORTED_NATIVEAOT_LIBRARY?

yes

lambdageek replied:
On second thought, I ended up making a reusable `add_imported_library_clr` function that just does the copy/strip/dependency handling and reimplemented `add_imported_nativeaot_library` on top of that.  `CLR_IMPORTED_COPY_TARGET` is now the property that the stripping logic looks for, and `CLR_IMPORTED_NATIVEAOT_LIBRARY` is just a marker for NativeAOT libs that we don't currently use for anything.

jkoritzinsky commented on src/native/managed/compile-native.proj:
```diff
@@ -0,0 +1,52 @@
+        <Message Importance="High" Text="Restoring native libs RuntimeConfiguration=$(RuntimeConfiguration) LibrariesConfiguration=$(LibrariesConfiguration)" />
+        <!-- workaround https://github.com/dotnet/msbuild/issues/2811 - Restore has to be a separate exectution with a different evaluation -->
+        <MSBuild Projects="@(NativeLibsProjectsToBuild)"
+                 Properties="Configuration=$(NativeLibsPublishConfiguration);RuntimeConfiguration=$(RuntimeConfiguration);LibrariesConfiguration=$(LibrariesConfiguration);RuntimeIdentifier=$(OutputRID);NativeLib=$(NativeLibKind)$(ExtraProps);_RandomPropertyToPleaseMSBuild=Restoring"
```
We use the `_MSBuildRestoreSessionId` property as the special one for restore in other places, can we use that here as well?

jkoritzinsky commented on src/native/managed/compile-native.proj:
```diff
@@ -0,0 +1,52 @@
+        <Message Importance="High" Text="Building native libs @(NativeLibsProjectsToBuild)" />
+        <Message Importance="High" Text="Building native libs RuntimeConfiguration=$(RuntimeConfiguration) LibrariesConfiguration=$(LibrariesConfiguration)" />
+        <MSBuild Projects="@(NativeLibsProjectsToBuild)"
+                 Properties="Configuration=$(NativeLibsPublishConfiguration);RuntimeConfiguration=$(RuntimeConfiguration);LibrariesConfiguration=$(LibrariesConfiguration);RuntimeIdentifier=$(OutputRID);NativeLib=$(NativeLibKind)$(ExtraProps);_RandomOtherPropertyToPleaseMSBuild=BuildingAndPublishing"
```
```suggestion
                 Properties="_IsPublishing=true;Configuration=$(NativeLibsPublishConfiguration);RuntimeConfiguration=$(RuntimeConfiguration);LibrariesConfiguration=$(LibrariesConfiguration);RuntimeIdentifier=$(OutputRID);NativeLib=$(NativeLibKind)$(ExtraProps);"
```

jkoritzinsky commented on src/native/managed/native-library.targets:
```diff
@@ -0,0 +1,84 @@
+
+    <Target Name="WriteNativeLibraryCMakeFragment" DependsOnTargets="SetupProperties;WriteNativeLibraryCMakeFragmentCommon;WriteNativeLibraryCMakeFragmentShared" AfterTargets="Publish" Condition="'$(PublishAot)' == 'true'"
+            Outputs="$(NativeLibraryCMakeFragmentPath);$(NativeLibraryCMakeFindPackageConfigPath)">
+        <!-- Note: GenerateFileFromTemplate uses ${replacementVar} syntax for things that it will replace.
+             Use a backtick to write a normal cmake variable reference: `${nonReplacedVar}
+             Wish: GenerateFileFromTemplate should provide a way to specify the replacement sigil.
+        -->
```
Open an issue for this to track improving it?

lambdageek replied:
https://github.com/dotnet/arcade/issues/14552

jkotas commented:
> windows stripping support

Why can't we just use the stripping done by the `dotnet publish`? It is what we do for `crossgen2` and `ilc` compilers that have the shipping binaries published using native AOT.

jkotas commented:
> `src/native/managed/libhellomanaged/libhellomanaged.csproj`
> `src/tools/Microsoft.DotNet.HelloManaged/src/Microsoft.DotNet.HelloManaged.csproj

What's the idea behing the library spread over multiple directories and multiple .csproj files like this? Why can't we just have a single `src/tools/Microsoft.DotNet.HelloManaged/src/Microsoft.DotNet.HelloManaged.csproj` that publishes the AOT library?

lambdageek commented:
> > windows stripping support
> 
> Why can't we just use the stripping done by the `dotnet publish`? It is what we do for `crossgen2` and `ilc` compilers that have the shipping binaries published using native AOT.

`dotnet publish` on Windows produces a `.dll` and a `.pdb` in the `artifacts/bin/libhellomanaged/.../publish` folder.  "windows stripping support" is about copying the .pdb to the right spot (artifacts/bin/coreclr/windows.x64.Debug/PDB/) along with the other runtime PDBs.
It's not doing any special stripping on Windows.

> What's the idea behing the library spread over multiple directories and multiple .csproj files like this? Why can't we just have a single src/tools/Microsoft.DotNet.HelloManaged/src/Microsoft.DotNet.HelloManaged.csproj that publishes the AOT library?

I'm not opposed.  I thought it would be easier to test or reuse the managed bits if they were in a completely separate assembly.  I wasn't sure how messy the .csproj changes would get and whether that would interfere with testing the managed bits in isolation.

lambdageek committed:
use _IsPublishing for the publish target

Co-authored-by: Jeremy Koritzinsky <jkoritzinsky@gmail.com>

lambdageek committed:
set MSBuildRestoreSessionId when restoring

lambdageek committed:
rename property to CLR_IMPORTED_COPY_TARGET

use it to decide whether we will have special strip logic

lambdageek committed:
add a general purpose add_imported_library_clr cmake function

it always strips and copies the symbols.

use it to implement add_imported_nativeaot_library

MichalStrehovsky commented:
> `dotnet publish` on Windows produces a `.dll` and a `.pdb` in the `artifacts/bin/libhellomanaged/.../publish` folder. "windows stripping support" is about copying the .pdb to the right spot (artifacts/bin/coreclr/windows.x64.Debug/PDB/) along with the other runtime PDBs. It's not doing any special stripping on Windows.

FWIW, placing symbols in the publish directory can be skipped with `CopyOutputSymbolsToPublishDirectory`. The symbols will still be available under `bin\native` (the output of native build).

jkotas commented:
Also, we deal with symbols in msbuild for the existing native AOT compiled components. We do not call back to cmake to move the symbols around.

lambdageek commented:
> Also, we deal with symbols in msbuild for the existing native AOT compiled components. We do not call back to cmake to move the symbols around.

I don't think the existing NativeAOT compiled components are a good model for what we're doing here.  All the logic about how the runtime subsystems are installed into `artifacts/bin/coreclr/<RID>/` are in CMake.  It's completely incidental whether the native binaries are compiled using clang or NativeAOT or rustc - CMake has all the logic about dependencies and linking the final outputs.  Moving that logic for some subcomponent into MSBuild would mean more duplication, not less.

am11 commented on src/native/managed/Templates/native-lib.find_package-config.cmake.in:
How do you feel about inlining these templates in MSBuild (to reduce number of moving parts)? It just something which we considered in part infra work. e.g. in cmake we need to use the config.in template files, it is a natural fit in that world. In MSBuild, what comes natural is to write bunch of lines from targets directly somewhere under obj/ directory; without introducing some external template system.

lambdageek replied:
@am11 an earlier version of this PR used `WriteLinesToFile` instead of `GenerateFileFromTemplate`.  I thought the template approach was cleaner.  But I could change it back.

am11 replied:
An alternative (to avoid external template and intermediate file writes) is to include `src/native/managed/**/` in some root-level CMakeLists via [GLOB_RECURSE](https://cmake.org/cmake/help/latest/command/file.html#glob-recurse), and pass values from msbuild to cmake via [this command-line args](https://github.com/dotnet/runtime/blob/fab69efde7d2458dfb23e01b686842cf6d7f576d/eng/build.sh#L566).

am11 replied:
An alternative (to avoid external template and intermediate file writes) is to include `src/native/managed/**/` in some root-level CMakeLists via [GLOB_RECURSE](https://cmake.org/cmake/help/latest/command/file.html#glob-recurse), and pass values of additional properties from msbuild to cmake via [this command-line args](https://github.com/dotnet/runtime/blob/fab69efde7d2458dfb23e01b686842cf6d7f576d/eng/build.sh#L566).

lambdageek commented:
Talked with @jkoritzinsky 

1. I will move the responsibility for placing the native library into `artifacts/coreclr/<rid>/...` from cmake to  MSBuild
2. Jeremy convinced me that the fact that cmake doesn't track the timestamp of imported shared libraries is not a problem - if we put the library in the right place.  If a C# change doesn't change the set of exports then there's no need to do any work in cmake.  If a C# change does change the exports or the signature of an export, then there should be a corresponding header file change, and cmake should pick that up and rebuild the C++ bits.

lambdageek committed:
Merge remote-tracking branch 'origin/main' into naot-runtime-lib

lambdageek committed:
move call to hellomanaged_Hello to CLRDataAccessCreateInstance

lambdageek committed:
WIP: copy runtime components to final location in native-library.targets

lambdageek committed:
add ::libs and ::headers imported lib targets

do not do any copying in cmake

lambdageek committed:
remove unused var

lambdageek committed:
add TODOs for new approach

lambdageek committed:
remove import-nativeaot-library.cmake

lambdageek committed:
remove add_imported_library_clr cmake function

lambdageek committed:
revert cmake changes for imported libs

remove the special install_clr and strip_symbols handling for imported libraries

lambdageek committed:
update README

lambdageek committed:
just emit one cmake file

lambdageek committed:
simplify the cmake template

lambdageek committed:
rename some cmake vars

lambdageek committed:
strip nativeoat-built componentlike coreclr does on mac/linux

windows remains to be done

lambdageek committed:
fixup osx and linux builds

jkoritzinsky reviewed (commented):
This is looking much better! I like the new approach!

lambdageek committed:
empty commit to trigger build

lambdageek committed:
copy windows PDBs to final install destinations

lambdageek committed:
DRY

lambdageek committed:
simplify example

lambdageek committed:
fixup comments and docs

lambdageek committed:
cleanup

lambdageek committed:
set library name on linux and mac

SharedLibraryInstallName only works on mobile apple platforms

lambdageek committed:
remove example

lambdageek commented:
I'm going to remove the example now.  I've seen it build with the latest changes and there's nothing interesting in the example itself anymore.  I think I'm ready for another round of reviews

lambdageek committed:
native-library.targets: make SetupOSSpecificProps probe for objcopy

lambdageek commented:
I think I want to remove the `libFoo::headers` imported target.  for the cdac work we end up using dlopen/LoadLibrary to get at the nativeaot'd lib, so there's not a build-time dependency between the shared library and the code that consumes it.  In fact it simplifies the build dependencies a  bit if locating the include files doesn't depend on AOTing the library - and also allows the crossdacs to consume the cdac reader (which is, by design, debuggee platform agnostic)

jkoritzinsky commented:
> I think I want to remove the `libFoo::headers` imported target. for the cdac work we end up using dlopen/LoadLibrary to get at the nativeaot'd lib, so there's not a build-time dependency between the shared library and the code that consumes it. In fact it simplifies the build dependencies a bit if locating the include files doesn't depend on AOTing the library - and also allows the crossdacs to consume the cdac reader (which is, by design, debuggee platform agnostic)

I think I agree here. If we exclusively reference the exports lazily through dlopen/LoadLibrary, then we don't need to have the header target.

If we were to link (dynamically or statically) against the lib, then I think we should expose the header as part of the same target that has the lib so linking against one CMake target would pull everything in needed to use that target.

In fact, if we are going to reference at runtime with dlopen/LoadLibrary, we don't need the full CMake integration at all. We only need one (imported?) target that adds the include directory for the header that has the function pointer types for dynamic interaction.

lambdageek committed:
remove support for exporting a header target

lambdageek commented:
> In fact, if we are going to reference at runtime with dlopen/LoadLibrary, we don't need the full CMake integration at all. 

yea, that's true.  I'll move the cmake stuff to a separate branch/PR and keep it on the back burner until we have a need for it.

> If we were to link (dynamically or statically) against the lib, then I think we should expose the header as part of the same target that has the lib so linking against one CMake target would pull everything in needed to use that target.

I'm not sure I totally agree.   Particularly for static linking it's sometimes helpful to only have a single target that depends on the libs, while multiple targets depend on the headers.  In any case, that can be hashed out later - when we have a need for the cmake stuff.

lambdageek committed:
Revert "remove support for exporting a header target"

This reverts commit 0f90a932a9cdbad8080e2bb465929072d3517487.

lambdageek committed:
Remove cmake integration

lambdageek commented:
Removed cmake integration.  Verified that the cdac stuff still builds.

lambdageek committed:
fix comment

jkoritzinsky commented on src/native/managed/compile-native.proj:
```diff
@@ -0,0 +1,67 @@
+                          AdditionalProperties="%(AdditionalProperties);@(SubprojectProps->'%(Identity)=%(Value)', ';')"/>
+    </ItemGroup>
+
+    <Target Name="BuildNativeLibs" BeforeTargets="Build" Condition="$(SupportsNativeAotComponents)">
+        <PropertyGroup>
+            <SplitSubprojectProps>@(SubprojectProps->'%(Identity)=%(Value)', ';')</SplitSubprojectProps>
+        </PropertyGroup>
+
+        <!-- workaround https://github.com/dotnet/msbuild/issues/2811 - Restore has to be a separate exectution with a different evaluation -->
+        <MSBuild Projects="@(NativeLibsProjectsToBuild)"
+                 Properties="MSBuildRestoreSessionId=$([System.Guid]::NewGuid());$(SplitSubprojectProps)"
+                 Targets="Restore" />
+
+        <MSBuild Projects="@(NativeLibsProjectsToBuild)"
+                 Properties="_IsPublishing=true;$(SplitSubprojectProps)"
+                 Targets="Build;Publish" />
+    </Target>
```
We should be able to reference the project just through a `ProjectReference`. Additionally, by making Publish the default target above, then we should automatically publish here when `runtime-prereqs.proj` references this through a ProjectReference (the build step would call the `Publish` target)

```suggestion
    <ItemGroup>
	      <ProjectReference Include="@(NativeLibsProjectsToBuild)" AdditionalProperties="$(SplitSubprojectProps)" Condition="$(SupportsNativeAotComponents)" />
    </ItemGroup>
```

jkoritzinsky commented on src/native/managed/compile-native.proj:
```diff
@@ -0,0 +1,67 @@
+<Project Sdk="Microsoft.Build.NoTargets">
+    <PropertyGroup>
+        <!-- We always want to use release for publishing using NativeAOT -->
+        <NativeLibsPublishConfiguration>Release</NativeLibsPublishConfiguration>
+	<!-- we always want to make shared libs -->
+        <NativeLibKind Condition="'$(NativeLibKind)' == ''">shared</NativeLibKind>
+    </PropertyGroup>
```
I think we can simplify this if we make it a traversal project.

```suggestion
<Project Sdk="Microsoft.Build.Traversal" DefaultTargets="Publish">
    <PropertyGroup>
        <!-- We always want to use release for publishing using NativeAOT -->
        <NativeLibsPublishConfiguration>Release</NativeLibsPublishConfiguration>
	<!-- we always want to make shared libs -->
        <NativeLibKind Condition="'$(NativeLibKind)' == ''">shared</NativeLibKind>
        <!-- When we publish, we want to ensure the SDK does the same thing as though we ran 'dotnet publish' -->
       <TraversalPublishGlobalProperties>$(TraversalPublishGlobalProperties);_IsPublishing=true</TraversalPublishGlobalProperties>
    </PropertyGroup>
```

lambdageek replied:
done.  thanks, this was a great idea!

jkoritzinsky reviewed (commented):
A few comments to reduce the MSBuild logic and reduce the number of steps to set up a new project.

lambdageek committed:
rename InstallRuntimeComponentDestination

lambdageek committed:
automatically include native-library.{props,targets}

for IsSourceProject C# projects under src/native/managed

lambdageek committed:
switch compile-native.proj to the traversal sdk

Co-Authored-By: Jeremy Koritzinsky <jekoritz@microsoft.com>

lambdageek commented on src/native/managed/README.md:
```diff
@@ -0,0 +1,35 @@
+
+In `src/native/managed/libMyNewLibrary/src/libMyNewLibrary.csproj`:
+1. Near the top,  add `<Import Project="..\..\native-library.props" />`
+2. Near the bottom, add `<Import Project="..\..\native-library.targets" />`
```
(I check for `'$(IsSourceProject)' == 'true' and '$(Languge)' == 'C#'` so that we can have managed tests, too.  but yes, this works)

jkoritzinsky reviewed (approved):
LGTM!
